Add credentials interface to Settings/Security

rcoll opened the pull request:
This PR is the beginning of the credentials interface for Jetpack Backups.

Fixes https://github.com/Automattic/wp-calypso/issues/18831

**Testing instructions:**
1. Select a self-hosted Jetpack site with a plan, navigate to settings/security, and attempt to add valid SSH, SFTP, or FTP credentials to the form. 
2. Ensure the credentials are present, and the site is connected, after refreshing and coming back to the form.
3. Ensure that credentials can be updated by resubmitting the form.

dmsnell commented on client/my-sites/site-settings/jetpack-credentials/credentials-form/index.jsx:
```diff
@@ -0,0 +1,280 @@
+										name="protocol"
+										value={ get( this.state.form, 'protocol', 'ssh' ) }
+										onChange={ this.handleFieldChange }
+										disabled={ credentialsUpdating }
```
this form doesn't really care about `credentialsUpdating` I don't think - I think it cares more about stuff like "form is submitting" or "form is disabled"

I mention this because as we start to step back and understand _what_ exactly we have, it becomes more general/reusable

dmsnell commented on client/my-sites/site-settings/jetpack-credentials/index.jsx:
```diff
@@ -0,0 +1,313 @@
+						: <Button primary onClick={ this.loadCredentialsForm }>{ translate( 'Ok, I understand' ) }</Button>
+				}
+			</CompactCard>
+		);
```
this here seems like a great candidate for extraction into a separate component

dmsnell commented on client/my-sites/site-settings/jetpack-credentials/index.jsx:
```diff
@@ -0,0 +1,313 @@
+					</h4>
+				</div>
+			</CompactCard>
+		);
```
this definitely looks like a good candidate to extract into a separate component

dmsnell commented on client/my-sites/site-settings/jetpack-credentials/index.jsx:
```diff
@@ -0,0 +1,313 @@
+			</div>
+		);
+	}
+}
```
notice how in this function we have several different places where we are checking for the state of the _machine_.

is it possible that we've discovered a _page_ or _section_ "wrapper" component which decodes the machine and display the appropriate _sub-page_ for each state?

it would surprise me if things didn't become simpler if we extracted these pages into separate pages and were left with one big dispatch…

```js
render() {
	const { setupStep } = this.props;

	switch ( setupStep ) {
		case 1:
			return <RewindStarter beginProcess={ this.nextStep } />;

		case 2:
			return <TOSConfirmation confirmTOS={ this.nextStep } />;

		case 3:
			return <CredentialsForm credentials={ … } submitCredentials={ this.submitCredentials } />;

		case 4:
			return <CredentialsList />;
	}

	return <UnknownStep />;
}
```

If we can continue to extract these things then we'll be left with much less to consider when we're working on any given piece of code, plus we can remove transition code we thought we needed. For example, we can handle all of the coordination in one function and spit out the Rewind state needed for this component - if we update the API, we can remove that translator.

```js
const getRewindState = ( state, siteId ) => {
	const rewindStatus = getRewindStatus( state, siteId );

	if ( rewindStatus === null ) {
		return { state: 'rewindDeactivated' };
	}

	const credentials = getCredentials( state, siteId );

	if ( ! credentials.length ) {
		return { state: 'waitingOnCredentials' };
	}

	return { state: 'rewindActive' }
}
```

^^^ that's not correct, but should convey the idea of containing and simplifying state as we get closer to the customer

roccotripaldi replied:
Well said Dennis. This section could be clarified a bit Rich.

For example, what gets rendered here if rewind is not active?

Both `renderConfiguredView` and `jetpack-credentials__header` hinge on the same logic. Could they be combined into a separate component, rather than an internal render method?

Both `CredentialsSetupFlow` and the popover bit hinge on the same logic, so they could be combined within the same component?

dmsnell commented:
@rcoll I left some comments on the main `render()` which is currently quite expansive. please consider the comment and see how much work you think it might take to continue our extraction. I think that at the point we have it extracted into the separate states of the Rewind activation we'll be able to remove some more of the data layer code and state tracking.

rcoll commented:
@dmsnell I went ahead and created a separate component for the setup flow, which greatly simplified `render()` in the main component. I also gave each step a meaningful name within the setup component. While the setup flow is still a single component, at least it removed from the main component and things seem much simpler and easier to follow now.

dmsnell commented on client/components/data/query-jetpack-credentials/index.jsx:
```diff
@@ -0,0 +1,32 @@
+}
+
+QueryJetpackCredentials.propTypes = {
+	requestCredentials: PropTypes.func,
```
seems like we can just make this mandatory - why do we have a default?

dmsnell commented on client/my-sites/site-settings/jetpack-credentials/credentials-form/index.jsx:
```diff
@@ -0,0 +1,261 @@
+
+		this.setState( {
+			form: { ...this.state.form, ...changedForm },
+			formError: { ...this.state.formErrors, resetErrors },
```
`...resetError` ?

dmsnell commented on client/my-sites/site-settings/jetpack-credentials/credentials-form/index.jsx:
```diff
@@ -0,0 +1,261 @@
+		const { formErrors } = this.state;
+
+		formState[ event.target.name ] = event.target.value;
+		formErrors[ event.target.name ] = false;
```
let's be careful not to directly mutate `this.state` !

rcoll replied:
Yep, this whole function actually wasn't even needed, so I removed it :-)

rcoll replied:
Yep, this whole part wasn't even needed, so I've removed it now.

dmsnell commented on client/my-sites/site-settings/jetpack-credentials/credentials-form/index.jsx:
```diff
@@ -0,0 +1,261 @@
+		if ( error ) {
+			this.setState( { formErrors: errors } );
+			return;
+		}
```
Note that we can also do this in a non-mutating way…

```js
const errors = Object.assign(
	! payload.host && { host: translate( … ) },
	! payload.port && { port: translate( … ) },
	isNaN( payload.port ) && { port: translate( … ) },
	…
);

return isEmpty( errors )
	? updateCredentials( siteId, payload )
	: this.setState( { formErrors: errors } );
```

dmsnell commented on client/my-sites/site-settings/jetpack-credentials/credentials-form/index.jsx:
```diff
@@ -0,0 +1,261 @@
+									) }
+								</FormLabel>
+							</td>
+						</tr>
```
After all this refactoring it may be more clear  how we can abstract some of this JSX…

what structure is shared among all the fields?
what information is different for each one?
how much of that different data is actually the same thing repeated multiple times (like a name)?

if we can get it to a few key properties then we can probably iterate of the descriptive data structure in a `map()`

```js
{ fields.map( ( { name, placeholder } ) => (
	<tr>
		…
		<FormTextInput { ...{
			name,
			placeholder: placeholder(), // needs to be like () => translate( … ) so translate gets called
			value: get( this.state.form, name, '' ),
			onChange={ this.handleFieldChange },
			isError: !! formErrors[ name ],
		} } />
	</tr>
) }
```

rcoll replied:
I tried this in the latest commit - it doesn't save a ton of code, but is nicer.

dmsnell commented on client/my-sites/site-settings/jetpack-credentials/credentials-setup-flow/index.jsx:
```diff
@@ -0,0 +1,139 @@
+				this.setState( { currentStep: 'form' } );
+				return;
+		}
+	};
```
Nice! Also fun fact, we have some ways to make this even more declarative - each step has a natural next step so let's just create that mapping…

```js
goToNextStep = () => this.setState( { currentStep: get( {
	start: 'tos',
	tos: 'form',
}, this.state.currentStep, this.state.currentStep ) } ); // default to no state transition
```

dmsnell commented on client/my-sites/site-settings/jetpack-credentials/credentials-setup-flow/index.jsx:
```diff
@@ -0,0 +1,139 @@
+			<div className="credentials-setup-flow">
+				{ 'start' === this.state.currentStep && this.renderStepStart() }
+				{ 'tos' === this.state.currentStep && this.renderStepTos() }
+				{ 'form' === this.state.currentStep && this.renderStepForm() }
```
Cool! This is a much nicer place to be than when we started. We're even ready to extract these into separate React components.

rcoll replied:
I think we can worry about additional react components after we get this merged. Thoughts?

dmsnell replied:
It's up to you, though I would ask probably "why not" here since the steps are already extracted into their own functions. In general we like to _either_ inline JSX inside of `render()` _or_ split them into separate components.

Here, we're _adding new code_ so the cost of doing it one way or the other is basically free in that we're not worried about introducing regressions to existing code. If we want to do it one way, let's just do it that way the first time because one of two things will happen: we'll never get back around to it and it will live on forever how we don't want it to; or we'll end up spending more time overall on it because when we come back to update it later (or _deal with it_) we won't be as efficient as we are now (when it's new and when we've been in it so much).

Now, you are free to decide that you don't want to do it the way I'm suggesting, and I will be fine with that. But, what I hope we don't do is push out new code that we think needs immediate refactoring - that's what PRs are for 😉

dmsnell commented:
Nice changes @rcoll - I left a few more comments and then I think that after we consider those it may be time again to reflect on the fetching code in the data layer and see if we can't remove a couple of those `_FAILURE` actions and associated tracking bits.

dmsnell commented on client/my-sites/site-settings/jetpack-credentials/credentials-setup-flow/index.jsx:
```diff
@@ -0,0 +1,142 @@
+
+	reset = () => {
+		this.setState( { currentStep: 'start' } );
+	};
```
fun tip!

we're free to eliminate the curly braces here 😄 

```js
reset = () => this.setState( { currentStep: 'start' } );
```

dmsnell commented on client/my-sites/site-settings/jetpack-credentials/credentials-form/index.jsx:
```diff
@@ -0,0 +1,226 @@
+									translate( 'Username' ),
+									translate( 'username' )
+								) }
+							</td>
```
If we only get two fields instead of all the fields out of it, we don't have to extract the JSX from here. When I first raised the idea I thought it might be possible to basically cover all the fields, `<tr>`, `<td>`, and others in such a way that it cuts out all of the manual bits.

If those bits remain _and_ we have a new function it may not be worth it, unless we can go all the way.

rcoll commented:
@roccotripaldi @dmsnell It took me some time to understand why/how to extract JSX as described, but I think I've got it now. Can you take a look and let me know if that's all correct? I think I've addressed all concerns listed to date, so please let me know how it looks now. Thanks :-)

dmsnell commented on client/my-sites/site-settings/jetpack-credentials/credentials-setup-flow/index.jsx:
```diff
@@ -0,0 +1,86 @@
+						updateCredentials={ updateCredentials }
+					/>
+				) }
+				<SetupFooter />
```
lovely!

dmsnell commented on client/my-sites/site-settings/jetpack-credentials/credentials-setup-flow/index.jsx:
```diff
@@ -0,0 +1,86 @@
+				{ 'start' === this.state.currentStep && (
+					<SetupStart
+						goToNextStep={ this.goToNextStep }
+						translate={ translate }
```
Note that if we don't need to pass `translate` down because we can wrap the sub-components in `localize()` just the same

dmsnell commented on client/my-sites/site-settings/jetpack-credentials/credentials-setup-flow/setup-footer.jsx:
```diff
@@ -0,0 +1,53 @@
+			<CompactCard className="credentials-setup-flow__footer">
+				<a
+					onClick={ this.togglePopover }
+					ref="popoverLink"
```
While this is okay, we generally prefer to use the _function syntax_ for `ref`s which keeps things a little tighter in the flow of data and control. Here's an example…

```js
// store the ref on some class property
storePopoverLink = ref => this.popoverLink = ref;

render() {
	…
	ref={ this.storePopoverLink }
	…
	content={ this.popoverLink }
	…
}
```

A couple things here:
	- `this.popoverLink` is _always_ valid; if the component hasn't yet mounted it will simply be `undefined`
	- no more "stringly-typed" references

dmsnell commented on client/my-sites/site-settings/jetpack-credentials/credentials-setup-flow/setup-footer.jsx:
```diff
@@ -0,0 +1,53 @@
+				</a>
+				<Popover
+					context={ this.refs && this.refs.popoverLink }
+					isVisible={ get( this.state, 'showPopover', false ) }
```
Not sure if we need `get()` here as if `this.state.showPopover` doesn't exist it will just be `undefined` and thus _falsey_. My guess is that it would evaluate properly for any expected values. `this.state` should always exist in these kinds of components (vs. functional stateless components)

alisterscott commented:
It looks like this may have broken refreshing or directly visiting the domains and plans pages. Raised as an issue here: https://github.com/Automattic/wp-calypso/issues/19262

seear commented:
This PR has also broken (or at least changed something to do with) auto-authorization when connecting a new Jetpack site from Calypso.

Before this PR (confirmed by resetting to before/after this commit) adding a new jetpack site from the _Add New Site_ button in Calypso would go through auth without needing any interaction. Now the flow stops at this point requiring a confirmation:

<img width="522" alt="screen shot 2017-10-30 at 17 19 22" src="https://user-images.githubusercontent.com/7767559/32185353-cbe549e0-bd96-11e7-8777-0a650e25b624.png">

Found by manually running this not-yet-merged [e2e test](https://github.com/Automattic/wp-e2e-tests/pull/750).

seear commented:
> This PR has also broken (or at least changed something to do with) auto-authorization when connecting a new Jetpack site from Calypso.

After investigation, it seems this PR is not at all related to display or non-display of the _Approve_ button during authorization. Confirmed by @rcoll.

Also seems highly unlikely that this PR is at all related to the e2e test failures previously [mentioned above](https://github.com/Automattic/wp-calypso/pull/18962#issuecomment-340324947).

Sorry for the wasted effort.

seear commented on client/state/jetpack/credentials/schema.js:
```diff
@@ -0,0 +1,14 @@
+			host: { type: 'string' },
+			port: { type: 'number' },
+			protocol: { type: 'string' },
+			pass: { type: 'bool' },
```
Wow, in the end this PR likely is causing those e2e tests to fail. This line breaks deserialization, meaning any e2e tests that (rightly or wrongly) rely on persisted state may fail.

Error:
```
calypso:state failed to load initial redux-store state +75ms TypeError: types[(t || "any")] is not a function
    at eval (VM8962 index.js:183)
    at Array.map (<anonymous>)
    at visit (VM8962 index.js:182)
    at eval (VM8962 index.js:532)
    at Array.forEach (<anonymous>)
    at visit (VM8962 index.js:529)
    at visit (VM8962 index.js:363)
    at compile (VM8962 index.js:546)
    at module.exports (VM8962 index.js:575)
    at isValidStateWithSchema (VM8953 utils.js:71)
```

Fix:
`pass: { type: 'boolean' },`

yoavf commented on client/my-sites/site-settings/jetpack-credentials/credentials-form/index.jsx:
```diff
@@ -0,0 +1,239 @@
+									<div>{ translate( 'Upload Path' ) }</div>
+									<FormTextInput
+										name="abspath"
+										placeholder={ translate( '/public_html' ) }
```
Let's "untranslate" this - no need for translation here.

yoavf commented on client/my-sites/site-settings/jetpack-credentials/credentials-form/index.jsx:
```diff
@@ -0,0 +1,239 @@
+										onChange={ this.handleFieldChange }
+										disabled={ formIsSubmitting }
+									>
+										<option value="ssh">{ translate( 'SSH' ) }</option>
```
Let's keep the protocols untranslated. Translated terms will just create confusion.

yoavf commented on client/my-sites/site-settings/jetpack-credentials/credentials-form/index.jsx:
```diff
@@ -0,0 +1,239 @@
+									<div>{ translate( 'Server Address' ) }</div>
+									<FormTextInput
+										name="host"
+										placeholder={ translate( 'yoursite.com' ) }
```
No need to translate this placeholder

yoavf commented on client/my-sites/site-settings/jetpack-credentials/credentials-form/index.jsx:
```diff
@@ -0,0 +1,239 @@
+									<div>{ translate( 'Port Number' ) }</div>
+									<FormTextInput
+										name="port"
+										placeholder={ translate( '22' ) }
```
This doesn't need translation
